src/backend/distributed/connection/connection_management.c

```diff
@@ -386,6 +345,28 @@ FindAvailableConnection(dlist_head *connections, uint32 flags)
 			continue;
 		}
 
+		if ((flags & NO_DATA_ACCESS_CONNECTION) != 0)
+		{
+			if (!connection->isNonDataAccess &&
+				ConnectionUsedForAnyPlacements(connection))
+			{
+				/* cannot use this connection for intermediate results */
+				continue;
+			}
+			else
+			{
+				connection->isNonDataAccess = true;
```

Comment:
Is there a reason not to mark  `isNonDataAccess`  after  calling `StartConnectionEstablishment()` instead of here? Otherwise, we might still end-up having more than one non-data access connections for certain CTE patterns?

I think it  could  also help simplifying this logic a bit, where `connection->isNonDataAccess == false`,  we don't need  to  check `ConnectionUsedForAnyPlacements(connection)`, because it'd be true by default.
Or, we could  assert `ConnectionUsedForAnyPlacements(connection) ==  false` when `connection->isNonDataAccess = true` etc.

(Plus, we should be able to remove `ClaimConnectionExclusively()` in  `RemoteFileDestReceiverStartup`  as long  as  we send  a  single intermediate result per node per query)

Reply:
Refactored it. I also found the double negation  (!isNon) confusing, so I renamed it to sideChannel.

Reply:
A downside I'm still looking to address is that we cannot reuse a cached connection.

Reply:
Introduced a connection purpose to distinguish between any (cached), data access and sidechannel.